lib/bundler/rubygems_integration.rb

```diff
@@ -620,296 +539,160 @@ def method_visibility(klass, method)
+      activated_spec_names = runtime.requested_specs.map(&:to_spec).sort_by(&:name)
+      [Gemdeps.new(runtime), activated_spec_names]
+    end
 
-      if Gem::Specification.respond_to?(:stubs_for)
-        def find_name(name)
-          Gem::Specification.stubs_for(name).map(&:to_spec)
-        end
-      else
-        def find_name(name)
-          Gem::Specification.stubs.find_all do |spec|
-            spec.name == name
-          end.map(&:to_spec)
-        end
+    if provides?(">= 2.5.2")
+      # RubyGems-generated binstubs call Kernel#gem
+      def binstubs_call_gem?
+        false
       end
 
-      if Gem::Specification.respond_to?(:default_stubs)
-        def default_stubs
-          Gem::Specification.default_stubs("*.gemspec")
-        end
-      else
-        def default_stubs
-          Gem::Specification.send(:default_stubs, "*.gemspec")
-        end
+      # only 2.5.2+ has all of the stub methods we want to use, and since this
+      # is a performance optimization _only_,
+      # we'll restrict ourselves to the most
+      # recent RG versions instead of all versions that have stubs
+      def stubs_provide_full_functionality?
+        true
       end
-
-      def use_gemdeps(gemfile)
-        ENV["BUNDLE_GEMFILE"] ||= File.expand_path(gemfile)
-        require "bundler/gemdeps"
-        runtime = Bundler.setup
-        Bundler.ui = nil
-        activated_spec_names = runtime.requested_specs.map(&:to_spec).sort_by(&:name)
-        [Gemdeps.new(runtime), activated_spec_names]
+    else
+      def binstubs_call_gem?
+        true
       end
 
-      if provides?(">= 2.5.2")
-        # RubyGems-generated binstubs call Kernel#gem
-        def binstubs_call_gem?
-          false
-        end
-
-        # only 2.5.2+ has all of the stub methods we want to use, and since this
-        # is a performance optimization _only_,
-        # we'll restrict ourselves to the most
-        # recent RG versions instead of all versions that have stubs
-        def stubs_provide_full_functionality?
-          true
-        end
+      def stubs_provide_full_functionality?
+        false
       end
     end
   end
 
   def self.rubygems
-    @rubygems ||= if RubygemsIntegration.provides?(">= 2.1.0")
-      RubygemsIntegration::MoreFuture.new
-    elsif RubygemsIntegration.provides?(">= 1.99.99")
-      RubygemsIntegration::Future.new
-    elsif RubygemsIntegration.provides?(">= 1.8.20")
-      RubygemsIntegration::MoreModern.new
-    elsif RubygemsIntegration.provides?(">= 1.8.5")
-      RubygemsIntegration::Modern.new
-    elsif RubygemsIntegration.provides?(">= 1.8.0")
-      RubygemsIntegration::AlmostModern.new
-    elsif RubygemsIntegration.provides?(">= 1.7.0")
-      RubygemsIntegration::Transitional.new
-    elsif RubygemsIntegration.provides?(">= 1.4.0")
-      RubygemsIntegration::Legacy.new
-    else # RubyGems 1.3.6 and 1.3.7
-      RubygemsIntegration::Ancient.new
-    end
+    @rubygems ||= RubygemsIntegration.new
```

Comment:
This is a good summary of this PR. Since currently `bundler` only supports rubygems 2.5 or higher, we can stick to the `MoreFuture` interface, which I unified in the `RubygemsIntegration` class.